resources/aws_ec2.go

```diff
@@ -663,3 +714,365 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error creating topic"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	log.Printf("%s: Created SNS Topic", obj)
+	// save this so we can delete the topic when we shut down
+	if topic.TopicArn != nil {
+		obj.topicArn = *topic.TopicArn
+	}
+}
+
+// makeCloudWatchRule makes a cloud watch rule to detect ec2 state changes
+func (obj *AwsEc2Res) makeCloudWatchRule() {
+	// make cloudwatch client
+	cweSess, err := session.NewSession(&aws.Config{
+		Region: aws.String("ca-central-1"),
+	})
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error creating session"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	obj.cweClient = cwe.New(cweSess)
+
+	// make cloudwatch rule
+	putRuleInput := &cwe.PutRuleInput{
+		Name:         aws.String(ruleName),
+		EventPattern: aws.String(eventPattern),
+	}
+	_, err = obj.cweClient.PutRule(putRuleInput)
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error creating session"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	log.Printf("%s: Created CloudWatch Rule", obj)
+}
+
+// targetCloudWatchRule configures cloudwatch to send events to sns topic
+func (obj *AwsEc2Res) targetCloudWatchRule() {
+	// target the rule to sns topic
+	target := &cwe.Target{
+		Arn:       aws.String(obj.topicArn),
+		Id:        aws.String(snsconst),
+		InputPath: aws.String(detail),
+	}
+	putTargetInput := &cwe.PutTargetsInput{
+		Rule:    aws.String(ruleName),
+		Targets: []*cwe.Target{target},
+	}
+	_, err := obj.cweClient.PutTargets(putTargetInput)
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error putting target"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	log.Printf("%s: Targeted SNS Topic", obj)
+}
+
+// snsSubscribe subscribes the endpoint to the sns topic
+func (obj *AwsEc2Res) snsSubscribe() {
+	// subscribe to the topic
+	subInput := &sns.SubscribeInput{
+		Endpoint: aws.String(obj.SnsEndpoint),
+		Protocol: aws.String(httpconst),
+		TopicArn: aws.String(obj.topicArn),
+	}
+	_, err := obj.snsClient.Subscribe(subInput)
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error subscribing to topic"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	log.Printf("%s: Created Subscription", obj)
+}
+
+// snsAuthorize adds the necessary permission for cloudwatch to publish to SNS
+func (obj *AwsEc2Res) snsAuthorize() {
+	gaInput := &sns.GetTopicAttributesInput{
+		TopicArn: aws.String(obj.topicArn),
+	}
+	attrs, err := obj.snsClient.GetTopicAttributes(gaInput)
+	if err != nil {
+		select {
+		case obj.awsChan <- &chanStruct{
+			err: errwrap.Wrapf(err, "error getting topic attributes"),
+		}:
+		case <-obj.closeChan:
+		}
+		return
+	}
+	// unmarshal the current sns publishing policy
+	type statement struct {
```

Comment:
are these structs already defined in the aws lib somewhere?

Reply:
Can't find these in in sns or aws. Not sure where else they could be.